site/content/rfds/0006-rust-plugin-sdk.md

```diff
@@ -0,0 +1,285 @@
+---
+title: Rust Plugin SDK
+weight: 6
+slug: 0006
+extra:
+  rfd: 6
+---
+
+Now that we've landed the initial implementation of the plugin system as
+described in [RFD 4], the next step is to split out our own existing
+analyses and data providers into separate plugins. We have already developed
+a couple of "dummy" plugins to test the proper functioning of the plugin
+system within `hc`, but the code in those plugins which implements the
+plugin gRPC interface and the query protocol is not well abstracted or
+reusable.
+
+Additionally, we'd like to start supporting folks outside of the Hipcheck
+project in developing their own plugins, and the same problem arises of that
+currently being difficult to do.
+
+Together, these problems motivate the need to begin developing Software
+Development Kits (SDKs) for building Hipcheck plugins.
+
+One of the strengths of the new plugin system is that plugins can be written
+in any programming language. At the same time, for building an SDK, we need
+to pick programming languages to prioritize. In our case, Rust is the obvious
+first choice since it's the language we use and the one we'll use for our
+own plugins. Over time we will no doubt add SDKs in other popular languages.
+
+The remainder of this RFD outlines the design of the Rust plugin SDK,
+particularly the interface we'll expose to plugin authors.
+
+## Plugin SDK Interface
+
+### The `Plugin` Trait
+
+The central trait of the plugin interface is the `Plugin` trait, which the
+user will use to define their plugin. The trait looks like this:
+
+```rust
+pub trait Plugin: Send + Sync + 'static {
+	/// The name of the publisher of the plugin.
+	const PUBLISHER: &'static str;
+
+	/// The name of the plugin.
+	const NAME: &'static str;
+
+	/// Handles setting configuration.
+	fn set_config(&mut self, config: JsonValue) -> StdResult<(), ConfigError>;
+
+	/// Gets the plugin's default policy expression.
+	fn default_policy_expr(&self) -> Result<String>;
+
+	/// Get all the queries supported by the plugin.
+	fn queries(&self) -> impl Iterator<Item = NamedQuery>;
+
+	/// Get the plugin's default query, if it has one.
+	fn default_query(&self) -> Option<DynQuery> {
+		self.queries()
+			.find_map(|named| named.is_default().then_some(named.inner))
+	}
+
+	/// Get all schemas for queries provided by the plugin.
+	fn schemas(&self) -> impl Iterator<Item = QuerySchema> {
+		self.queries().map(|query| QuerySchema {
+			query_name: query.name,
+			input_schema: query.inner.input_schema(),
+			output_schema: query.inner.output_schema(),
+		})
+	}
```

Comment:
This implementation doesn't address the "default query explanation" gRPC call we added during the RFD4 impl

Reply:
Good catch. I'll add that in.